include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -800,6 +880,74 @@ namespace VectorTools
     }
 
 
+
+    /**
+     * Generic implementation for the project() function on distributed triangulations.
+     */
+    template <int dim, int spacedim, typename VectorType, int components, int fe_degree>
+    void do_project_distributed (const Mapping<dim, spacedim>                              &mapping,
+                                 const DoFHandler<dim, spacedim>                           &dof,
+                                 const ConstraintMatrix                                    &constraints,
+                                 const Quadrature<dim>                                     &quadrature,
+                                 const Function<spacedim, typename VectorType::value_type> &function,
+                                 VectorType                                                &vec_result,
+                                 const bool                                                 enforce_zero_boundary,
+                                 const Quadrature<dim-1>                                   &q_boundary,
+                                 const bool                                                 project_to_boundary_first)
+    {
+      const parallel::distributed::Triangulation<dim,spacedim> *parallel_tria =
```

Comment:
@kronbichler correct, i was only thinking about assembly part, not the matrix-free solver itself.

@masterleinad :  i would still cast to `p::Tria` but would wait with `p::s::Tria` unit tests until we known that MatrixFree works with it.